### DISCUSSION TITLE
Styling methods

### DISCUSSION THREAD

**Discussion by ed2050** [zauberzeug/nicegui#1806] [Ideas / Feature Requests]
Continuing from discussion #1778 -
nicegui has too many styling methods.  Hiding the tailwind/quasar/css complexity is good, but then you expose it again by having 3 overlapping methods: style, props, and classes.  From user perspective those are all slightly different ways to do the same thing, control styling.  It's confusing to keep track of which one to apply in which situation, and requires knowledge of tailwind / quasar classes and properties.  As I code, I'm always asking myself "ok I remember w-full makes an element full width, but is it tailwind or quasar?  Do I call .style ('w-full') or .props ('w-full') or .classes ('w-full')"?

# Option 1
It would be better to have one method like `.style()` and then disambiguate on the back end whether the argument is a css style rule, a tailwind class, or a quasar/html property.  Nicegui implementation would be more difficult, but much easier for users to have one consistent method.

Ideally the implementation would have a single argument with consistent structure.  Values should use the same separator.  To help differentiate css style rules from other types, it can use a unique separator like `/` between elements:
Eg. `element.style (styling = "color : blue ; / text-h4 / width=25vw")`

To parse, nicegui would split on / and determine whether each item is a css style rule, a tailwind class (can be tested with list of all known tailwind classes), a quasar property (same?), or an arbitrary class identifier.  There may be overlap between tailwind and quasar values, which can be handled by either 1) giving precedence to one over the other, or 2) adding a disambiguating flag like t: for tailwind and q: for quasar.  However this can get messy and complicated, especially if adding a new flag that has be parsed and removed from the string.

Also, parsing may be complicated by the fact that css [can allow any character](https://www.w3.org/TR/CSS21/syndata.html#:~:text=In%20CSS%2C%20identifiers%20(including%20element,hyphen%20followed%20by%20a%20digit.) escaped with `\`.  So even / can appear in a css style rule.  Nicegui could either disallow / in style rules (it's a pretty rare case anyway) or use a more complicated parser to ignore / within style rules.  / is just an example, same applies to any separator character.

The advantage of this approach is that the user only has to remember one method `.style()` and pass whatever style args they want to it.

# Option 2
Alternatively, if the tradeoffs for parsing one argument are too complicated, then I suggest following the python maxim "explicit is better than implicit".  Have one method `.style ()` with three optional parameters:
* `css = ` for css style rules
* `tailwind =` for tailwind classes
* `quasar = ` for quasar properties

That makes it very clear which arguments belongs where.  Users only have to remember one method, which is an improvement.  They still have to distinguish tailwind from quasar from css, but at least the code makes it obvious which one they're using.  And no parsing on the backend, implementation is trivial.   Depending on the flag, just call the correct .props or .classes method (which can be changed to private methods).

The downside is this approach still requires the user to differentiate between tailwind and quasar items.  But that may be necessary anyway as long as nicegui accepts tailwind / quasar values.  To have complete abstraction, nicegui would need to create its own style language like `width = 50%` and then translate that into tailwind or quasar settings as appropriate.  I don't think that level of abstraction is necessary or desirable.  Users should have access to the underlying power of tailwind and quasar if needed.  They just shouldn't be forced to use it for normal cases.

# Conclusion
I think option 2 may be the better approach.  Option 1 is simpler to learn and use, but may make code harder to read (is `foobar` a tailwind class or a quasar property or a css class?  where do I find it?) and definitely complicates the backend implementation.  Option 2 is very readable and should be easy to implement.

**Comment by natankeddem** [comment]
I personally don't see the problem with the current way things work. Yes, it takes some effort to learn how the systems that feed into styling in NiceGUI, but it isn't horrible in my opinion. I don't see any benefit of having these alternate mechanisms for applying props/classes/css to elements would benefit my current workflow.

More recently I have tried to limit usage to tailwind through the `tailwind` property and I find it works really well with the autocomplete to get me nicely self documenting tailwind classes. I only utilize classes for anything that isn't in tailwind. The only improvement I think could be useful, but a ton of work would be to make all quasar elements expose props via class properties. That way auto-complete would work and you wouldn't always have to refer back to the quasar documentation as much hopefully.

Maybe there could be a similar scheme to the `tailwind` property to get auto-complete for css as well? Then you could do:
```py
ui.label("hi").css.color("red")
```

**Reply by natankeddem** [reply]
Thinking more on this, `style` could possibly be refactored into a class. Then it could contain it's current functionality as well as provide the new `tailwind` like functionality.

**Reply by falkoschindler** [reply]
Good idea, @natankeddem. I've already thought about such a `css` class. But I couldn't find any (easily parsable) CSS reference which we could transform into a Python module. And even if there was: If the module gets too big, the auto-completion starts to struggle. I already had to cut some corners for Tailwind.

**Reply by natankeddem** [reply]
@falkoschindler I did find this:
https://github.com/mdn/data/tree/main/css
The `.json` files there should contain all/most of the information theoretically I think. Although, as you stated `style` isn't that utilized, so maybe not worth the effort.

**Comment by falkoschindler** [comment]
Thanks for the suggestions, @ed2050!

I have some concerns:
- I think we shouldn't mix all three methods (option 1). They have different meaning, even though all three _can_ control styling. But props often control behavior. Mixing them with a new syntax complicates the parser and - in my opinion - the usage. Props could easily contain slashes, and simply CSS [color](https://developer.mozilla.org/en-US/docs/Web/CSS/color_value) values can as well. Even without slashes its tricky to communicate how to use such a method.
- One `style` method with `css=`, `tailwind=` and `quasar=` parameters only shifts the problem, I think. And it collides with `add=`, `update=` and `remove=` parameters. For, e.g., removing a class, we would need a different solution. And sometimes you need a non-Tailwind class or a non-Quasar prop; how to do that?

In summary I would continue following this strategy:
- Try to reduce the need for using `.props` by introducing parameters like `ui.button(..., icon=...)`. This isn't always easy (e.g. button color), but we could spent more effort in common props like "flat", "outlined", "fab", "size", ...
- Try to reduce the need for using `.style`. This is already working quite well. In my day-to-day work I rarely need this method.
- Support auto-completion for Tailwind classes. This has been achieved using the `.tailwind` attribute.
- Improve the documentation. All other effort is pointless if the user doesn't know about how to use it.

**Comment by ed2050** [comment]
> Try to reduce the need for using .props by introducing parameters like ui.button(..., icon=...). This isn't always easy (e.g. button color), but we could spent more effort in common props like "flat", "outlined", "fab", "size", ...

This approach scales poorly.  Adding more params for style properties makes code harder to read, write, update.  It mixes functional elements with style elements.  Your code will be littered with params like size= , outlined=, etc.  Any sizable app will be a nightmare to update the appearance of.

It's better to separate styling from functionality.  Think MVC paradigm for gui, you don't mix the data with the view with the controller.  Think css and html.  No one wants to update a thousand width= properties throughout their source pages when they can edit a single "button { width: 5em }" in css.

If people don't want a single method for styling, ok.  Having .classes .style and .props adds complexity, but I can live it.  At least have a way to apply those methods away from the functional structure.

One way to support this would be selective iterators over the element tree.  Say your root element is `root`.  

- You could have `root.buttons ()` to iterate over all buttons under root, or `root.labels ()` to iterate over all labels, regardless of where in the tree they are.
- Even more useful would be [css selectors like beautifulsoup](https://www.crummy.com/software/BeautifulSoup/bs4/doc/#css-selectors-through-the-css-property),  so you can do something like `root.select ('button.download')` to get all buttons with class download, or `root.select ('button:first-child')` to get the first button in every group.

For example:

```py
# purely functional declarations, easy to see structure
def add_nodes (root) :
    with root :
        ui.label ('Are you ready?')
        ui.button ('Let's go!')
    return root

# all styling applied here
def add_styles (root) :
    for label in root.labels () :   # return all labels under root?
        label.classes ('bg-blue-200')

   for button in root.buttons () :   # return all button under root?
       button.classes ('w-25 bg-green-500')
    return root

def main () :
    root = ui.column ()
    add_nodes (root)
    add_styles (root)
```

# Advantages
- Cleaner code that's easier to read.  Don't have to wade through nonfunctional cruft like `ui.card ().classes ('bg-blue-400 dark:bg-blue-600').props ('w-52')` on every line.
- Much easier to update styling when it's all in one place.  Can have separate functions or modules that apply styling to the element tree.  Finding all the places buttons are set to a certain width or color is much easier when styling code is grouped together.
- Easier to see related issues.  If I make this image width 75%, what other elements do I need to adjust widths on to compensate?  When styles are mixed with functionality, those elements can be scattered all over your source in different functions and different pages.  Imagine having a dozen @ui.page functions across multiple python modules.
- Easier to change styling.  Can be as simple as calling a different function: `apply_styles (root)` can be easily changed to `apply_alt_styles (root)`.  This not only lets you test appearance changes easily, but you can effortlessly switch appearance on the fly based on user settings (think slider control for light or dark mode, but affecting any aspect of appearance not just colors).

For simple apps, sure none of this matters.  When all your code is <250 lines, everything is easy to find.  

As app complexity grows, mixing functionality and appearance becomes a real problem.  I've worked on python code bases from 10k to severel hundred k lines of code.  Things can very quickly spiral out of control when separation is lacking.

**Reply by natankeddem** [reply]
Have you looked at the `default_props`, `default_classes`, `default_style` methods? those will allow you to control styling on multiple elements at once centrally, but you will need to re-instantiate them after the changes. You can also use these in conjunction with extending the element classes to get different outcomes.
I have also been playing around recently with utilizing mostly extended elements in a recent project; which package all my styling internally. I keep my extended element classes in a module together `elements.py` which all the other modules utilize. This gives me a central place to look at the majority of my styling in the project.

**Reply by falkoschindler** [reply]
> I have also been playing around recently with utilizing mostly extended elements in a recent project; which package all my styling internally.

That's what I usually do as well. You have the full power of the Python language at your fingertips. Subclassing NiceGUI's elements to more customized building blocks of your app works really well. And actually you don't even need to subclass. Wrapper functions work as well.

Default props/classes/style might also help. But if you have different, e.g., buttons in your app - say primary, secondary, destructive, ... - you might not want to apply the same styling to `ui.button`.

And sometimes I also have some style.css where I collect CSS definitions with proper syntax highlighting and auto-completion. In Python I simply load it into `ui.add_head_html`. Sure, that's more a web-dev approach and not very pythonic, but nevertheless it's one of many solutions to the problem.

**Reply by ed2050** [reply]
Thanks for the suggestions, @falkoschindler and @natankeddem .  These are useful.

> Have you looked at the default_props, default_classes, default_style methods?

No, are those in the docs somewhere?  Couldn't find anything on main docs page, maybe it's on a subpage somewhere?

> Default props/classes/style might also help. But if you have different, e.g., buttons in your app - say primary, secondary, destructive, ... - you might not want to apply the same styling to ui.button.

Exactly.  Default styles are necessary but not sufficient.

> >I have also been playing around recently with utilizing mostly extended elements in a recent project; which package all my styling internally.

> That's what I usually do as well. You have the full power of the Python language at your fingertips. Subclassing NiceGUI's elements to more customized building blocks of your app works really well. And actually you don't even need to subclass. Wrapper functions work as well.

Do you have examples of this?  Especially wrappers.

> And sometimes I also have some style.css where I collect CSS definitions with proper syntax highlighting and auto-completion. In Python I simply load it into ui.add_head_html. Sure, that's more a web-dev approach and not very pythonic, but nevertheless it's one of many solutions to the problem.

It's a very useful approach that I've been doing.  Unfortunately you lose most of the tailwind / quasar niceness this way.  If you're just changing colors and font styling, it's great.  Otherwise, making sure your css is responsive and plays nice with tailwind / quasar is time consuming.

Speaking of which: ui.add_head_html puts the element before tailwind (and quasar) stylesheets in the head.  Meaning tailwind overrides any settings that conflict.  So you either use `!important`, which causes other problems.  Or what I do is `ui.add_html (f'<style> { mystyles } </style>')` in the doc body.  Put it in the footer to make sure it's the last style applied.

**Reply by natankeddem** [reply]
The default methods are documented here:
https://nicegui.io/documentation/element
Also, the tests might give you some ideas as well:
https://github.com/zauberzeug/nicegui/blob/main/tests/test_element.py
In a minimal application you might be able to get away with just using these alone, but in more elaborate projects you might need to use them in conjunction with class extension.

I should be able to link you to a new project which I am working on that is using some of these ideas in the next couple of weeks as well.

**Reply by falkoschindler** [reply]
@ed2050
> Do you have examples of this? Especially wrappers.

Here is a tiny example: https://github.com/zauberzeug/nicegui/discussions/1834#discussioncomment-7322673

**Reply by natankeddem** [reply]
I guess it can't hurt to show off my work in progress element module. Some of these changes are styling but some of the extensions change functionality or even argument types. For example I have some `ui.spinners` that I want to group together visually; they can all registers a `master` `ui.spinner` to become visible all at once. Also, I typically want the spinners to start in an invisible state and only turn on once I am showing progress. These are requirements for my specific project so I baked it into these new elements.
```py
from typing import Any, Callable, Dict, List, Optional, Union
from nicegui import ui, Tailwind
from nicegui.elements.spinner import SpinnerTypes
from nicegui.tailwind_types.height import Height
from nicegui.tailwind_types.width import Width
from nicegui.elements.mixins.validation_element import ValidationElement
import logging

logger = logging.getLogger(__name__)
orange = "#f59e0b"
dark = "#171717"
ui.card.default_style("max-width: none")
ui.card.default_props("flat bordered")
ui.input.default_props("outlined dense hide-bottom-space")
ui.button.default_props("outline dense")
ui.select.default_props("outlined dense dense-options")
ui.checkbox.default_props("dense")


class ErrorAggregator:
    def __init__(self, *elements: ValidationElement) -> None:
        self.elements = list(elements)

    def clear(self):
        self.elements.clear()

    def append(self, element: ValidationElement):
        self.elements.append(element)

    def remove(self, element: ValidationElement):
        self.elements.remove(element)

    @property
    def no_errors(self) -> bool:
        return all(validation(element.value) for element in self.elements for validation in element.validation.values())


class WColumn(ui.column):
    def __init__(self) -> None:
        super().__init__()
        self.tailwind.width("full").align_items("center")


class DBody(ui.column):
    def __init__(self, height: Height = "[480px]", width: Width = "[240px]") -> None:
        super().__init__()
        self.tailwind.align_items("center").justify_content("between")
        self.tailwind.height(height).width(width)


class WRow(ui.row):
    def __init__(self) -> None:
        super().__init__()
        self.tailwind.width("full").align_items("center").justify_content("center")


class Card(ui.card):
    def __init__(self) -> None:
        super().__init__()
        self.tailwind.border_color("primary")


class DInput(ui.input):
    def __init__(
        self,
        label: str | None = None,
        *,
        placeholder: str | None = None,
        value: str = " ",
        password: bool = False,
        password_toggle_button: bool = False,
        on_change: Callable[..., Any] | None = None,
        autocomplete: List[str] | None = None,
        validation: Callable[..., Any] = bool,
    ) -> None:
        super().__init__(
            label,
            placeholder=placeholder,
            value=value,
            password=password,
            password_toggle_button=password_toggle_button,
            on_change=on_change,
            autocomplete=autocomplete,
            validation={"": validation},
        )
        self.tailwind.width("full")
        if value == " ":
            self.value = ""


class FInput(ui.input):
    def __init__(
        self,
        label: str | None = None,
        *,
        placeholder: str | None = None,
        value: str = " ",
        password: bool = False,
        password_toggle_button: bool = False,
        on_change: Callable[..., Any] | None = None,
        autocomplete: List[str] | None = None,
        validation: Callable[..., Any] = bool,
        read_only: bool = False,
    ) -> None:
        super().__init__(
            label,
            placeholder=placeholder,
            value=value,
            password=password,
            password_toggle_button=password_toggle_button,
            on_change=on_change,
            autocomplete=autocomplete,
            validation={} if read_only else {"": validation},
        )
        self.tailwind.width("64")
        if value == " ":
            self.value = ""
        if read_only:
            self.props("readonly")


class DSelect(ui.select):
    def __init__(
        self,
        options: List | Dict,
        *,
        label: str | None = None,
        value: Any = None,
        on_change: Callable[..., Any] | None = None,
        with_input: bool = False,
        multiple: bool = False,
        clearable: bool = False,
    ) -> None:
        super().__init__(
            options, label=label, value=value, on_change=on_change, with_input=with_input, multiple=multiple, clearable=clearable
        )
        self.tailwind.width("full").max_height("[40px]")


class FSelect(ui.select):
    def __init__(
        self,
        options: List | Dict,
        *,
        label: str | None = None,
        value: Any = None,
        on_change: Callable[..., Any] | None = None,
        with_input: bool = False,
        multiple: bool = False,
        clearable: bool = False,
    ) -> None:
        super().__init__(
            options, label=label, value=value, on_change=on_change, with_input=with_input, multiple=multiple, clearable=clearable
        )
        self.tailwind.width("64")


class DButton(ui.button):
    def __init__(
        self,
        text: str = "",
        *,
        on_click: Callable[..., Any] | None = None,
        color: Optional[str] = "primary",
        icon: str | None = None,
    ) -> None:
        super().__init__(text, on_click=on_click, color=color, icon=icon)
        self.props("size=md")
        self.tailwind.padding("px-2.5").padding("py-1")


class DCheckbox(ui.checkbox):
    def __init__(self, text: str = "", *, value: bool = False, on_change: Callable[..., Any] | None = None) -> None:
        super().__init__(text, value=value, on_change=on_change)
        self.tailwind.width("full").text_color("secondary")


class IButton(ui.button):
    def __init__(
        self,
        text: str = "",
        *,
        on_click: Callable[..., Any] | None = None,
        color: Optional[str] = "primary",
        icon: str | None = None,
    ) -> None:
        super().__init__(text, on_click=on_click, color=color, icon=icon)
        self.props("size=sm")


class SmButton(ui.button):
    def __init__(
        self,
        text: str = "",
        *,
        on_click: Callable[..., Any] | None = None,
        color: Optional[str] = "primary",
        icon: str | None = None,
    ) -> None:
        super().__init__(text, on_click=on_click, color=color, icon=icon)
        self.props("size=sm")
        self.tailwind.width("16")


class LgButton(ui.button):
    def __init__(
        self,
        text: str = "",
        *,
        on_click: Callable[..., Any] | None = None,
        color: Optional[str] = "primary",
        icon: str | None = None,
    ) -> None:
        super().__init__(text, on_click=on_click, color=color, icon=icon)
        self.props("size=md")


class Spinner(ui.spinner):
    def __init__(
        self,
        type: SpinnerTypes | None = "bars",
        *,
        size: str = "lg",
        color: str | None = "primary",
        thickness: float = 5,
        master: ui.spinner | None = None,
    ) -> None:
        super().__init__(type, size=size, color=color, thickness=thickness)
        self.visible = False
        if master is not None:
            self.bind_visibility_from(master, "visible")
```
